### PR TITLE
Follow tombstone and predecessor history

### PR DESCRIPTION
**Pull request by MadLittleMods**
Follow tombstone and predecessor history

Fix https://github.com/matrix-org/matrix-public-archive/issues/59

Other updates:

 - Update tests to use `/roomid/room1/date/2022/01/03` format instead of trying to retrofit the weird alias stuff on there. Which also makes the fancy to actual URL utilities much more simple.
 - Update to specify `archiveMessageLimit` in the test case because pages have different number of events depending on if we are against a boundary, hidden events, etc.

### Dev notes

 - [x] Follow tombstone and predecessor history when visiting room on a given day
 - [x] Follow tombstone and predecessor history for `/jump`
 - [x] Write tests
 - [ ] ~~Edge case: Someone can create a circular reference which will cause us to keep redirecting the user around in a loop. Do we care?~~
    - Maybe not as the `ERR_TOO_MANY_REDIRECTS` in the browser will handle it
 - [x] Does it work in a Gitter-like historical room where we created the room with some date X and imported history with timestamps before X?
    - Do we need some causality query parameter for `/timestamp_to_event`? Like `/timestamp_to_event?dir=f&ts=123&event_id=$123` meaning find the closest event from `ts` looking forward after `event_id` (the causality part, "A after B")
       - Perhaps we could just see some causality by observing `m.room.create` popping up in `/timestamp_to_event?dir=f` when we `/jump`. If we see `m.room.create`, then assume we've reached the end of the room and move to the successor
       - Someone could also tie us in knots by making timestamps circular so we probably do need a causality param for `/timestamp_to_event`
       - Created [MSC3999](https://github.com/matrix-org/matrix-spec-proposals/pull/3999) to track adding a causal query parameter to `/timestamp_to_event?event_id=$abc`

### CODE REVIEW

File: server/lib/matrix-utils/ensure-room-joined.js [hunk 1]
```diff
@@ -4,15 +4,29 @@ const assert = require('assert');
...
+
+  // Let's do our best for the user to join the room. Since room ID's are
+  // unroutable on their own and won't be found if the server doesn't already
+  // know about the room, we'll try to join the room via the server name that
+  // we derived from the room ID or alias.
+  const viaServersWithAssumptions = new Set(viaServers);
+  const derivedServerName = getServerNameFromMatrixRoomIdOrAlias(roomIdOrAlias);
+  if (derivedServerName) {
+    viaServersWithAssumptions.add(derivedServerName);
+  }
```
Comment (MadLittleMods) [inline server/lib/matrix-utils/ensure-room-joined.js:26]: TODO: Split this off into a separate PR

File: server/routes/room-routes.js [hunk 2]
```diff
@@ -265,39 +282,84 @@ router.get(
...
-      // in the displayed range, say `T12:00:05` would still give us the same day
-      // `/2020/01/02` and we want to redirect them to previous chunk from that same
-      // day, like `/2020/01/02T12:00:00`
+      // in the displayed range (say that occured on `T12:00:25`) would still give us
+      // the same day `/2020/01/02` and we want to redirect them to previous chunk from
+      // that same day that still encompasses the closest message looking backwards,
+      // like `/2020/01/02T13:00:00`
       if (dir === DIRECTION.backward) {
+        // We choose `currentRangeStartTs` instead of `ts` (the jump point) because TODO: why?
+        // and we don't choose `currentRangeEndTs` because TODO: why?
```
Comment (MadLittleMods) [inline server/routes/room-routes.js:352]: I feel like I can't justify this 🤔 but it does seem correct from tested behavior.

I think my reasoning for using `currentRangeEndTs` before was that we could keep paginating with day precision if we were showing more than a day for each page. But then this falls apart when you have a bunch of  messages from the same hour that fill up multiple pages.

Using `ts` makes various tests fail with various precision differences. Probably because of the `- 1`/`+ 1` stuff that we want to remove in the future anyway when MSC3999 lands.

But those don't seem like explanations for why (the truth kernel)

Comment (MadLittleMods) [reply server/routes/room-routes.js]: I've left a link to this discussion thread so people can get some context. But I think I will opt to merging this PR without figuring this out. Will just have to contemplate this more and fill it out when an epiphany happens.
